### PR TITLE
Adding tangent tensor calculation and refactoring of the FemDem elements

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm improving the quality of the code in my app and I've added the tangent tensor calculation in my app too.

### CODE REVIEW

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp [hunk 1]
```diff
@@ -38,12 +38,10 @@ FemDem3DElement::FemDem3DElement(IndexType NewId, GeometryType::Pointer pGeometr
 
 	// Each component == Each edge
 	mNumberOfEdges = 6;
-	mF_sigmas = ZeroVector(mNumberOfEdges);   // Equivalent stress
+	mNonConvergedThresholds = ZeroVector(mNumberOfEdges);   // Equivalent stress
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_element.cpp:41]: Most of the code in 2d and 3d is duplicated, can you simplify that?

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp [hunk 2]
```diff
@@ -1010,97 +1030,101 @@ double FemDem3DElement::GetMaxValue(Vector Strain)
...
+	for (std::size_t i = 1; i < non_zero_values.size(); ++i) {
+		if (non_zero_values[i] > aux)
+			aux = non_zero_values[i];
 	}
-	return V[n - 1];
+	return aux;
 }
 
-double FemDem3DElement::GetMinAbsValue(Vector Strain)
+double FemDem3DElement::GetMinAbsValue(
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_element.cpp:1052]: This is not very efficient, look at the StructuralMechanics implementation

File: applications/MeshingApplication/custom_processes/mmg_process.cpp [hunk 3]
```diff
@@ -212,6 +212,23 @@ void MmgProcess<TMMGLibray>::ExecuteInitializeSolutionStep()
     // Save to file
     if (safe_to_file) SaveSolutionToFile(false);
 
+    // We do a clear of the containers data (this is not interpolated anyway and some variables give problems, because are pointers to others containers)
```
Comment (loumalouomega) [inline applications/MeshingApplication/custom_processes/mmg_process.cpp]: Remove these changes from this PR

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Can you unify the 2D/3D element?, can you look to my others comments?

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp:1052]: do you mean the one implemented in the tangent calculator?

thx

Comment (loumalouomega) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp:1052]: Yep

File: applications/MeshingApplication/custom_processes/mmg_process.cpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/MeshingApplication/custom_processes/mmg_process.cpp]: done

**Comment by loumalouomega** [comment]
Can you unify the 2D/3D element in one file¿

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp (hunk 1 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp:41]: I know that but some parts of the algorithm are slightly differetn in 2d and 3d so it's not possible. In another PR I can create a library of yield surfaces like in structural in order to simplify the elements...

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp:1052]: done, thx

**Review by loumalouomega — DISMISSED** [review]
OK

**Review by loumalouomega — DISMISSED** [review]

**Comment by AlejandroCornejo** [comment]
@loumalouomega could you reapprove? thx

File: applications/FemToDemApplication/custom_elements/femdem2d_element.cpp [hunk 4]
```diff
@@ -351,7 +351,7 @@ void FemDem2DElement::CalculateLocalSystem(MatrixType &rLeftHandSideMatrix, Vect
 		const Matrix& C =  Values.GetConstitutiveMatrix();
 
 		Matrix tangent_tensor;
-		if (is_damaging == true) {
+		if (is_damaging == true && std::abs(strain_vector[0] + strain_vector[0] +strain_vector[0]) > tolerance) {
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: [0]...[0]...[0]

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp [hunk 5]
```diff
@@ -462,7 +462,7 @@ void FemDem3DElement::CalculateLocalSystem(
 		this->CalculateDeformationMatrix(B, DN_DX);
 
 		Matrix tangent_tensor;
-		if (is_damaging == true) {
+		if (is_damaging == true && std::abs(StrainVector[0] + StrainVector[1]) > tolerance) {
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: Here only two?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Look my commments

File: applications/FemToDemApplication/custom_elements/femdem2d_element.cpp (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: LOL you're rigth TT thank you!!

Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: done! you're a wizard

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp (hunk 5 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: done too

File: applications/FemToDemApplication/custom_elements/femdem2d_element.cpp (hunk 4 above)
Comment (loumalouomega) [reply applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: <3

**Review by loumalouomega — APPROVED** [review]
Fine

**Pull request merged**